src/GlobImports/Exe.hs

```diff
@@ -78,80 +80,118 @@
+import System.FilePath.Glob
+import System.Process.Typed
 
 -- | The source file location. This is the first argument passed to the
 -- preprocessor.
-newtype Source = Source { unSource :: FilePath }
+newtype Source = Source {unSource :: FilePath}
 
 -- | The source file contents. This is the 'String' contained in the file of the
 -- second argument passed to the preprocessor.
-newtype SourceContents = SourceContents { unSourceContents :: String }
+newtype SourceContents = SourceContents {unSourceContents :: String}
 
 -- | The destination file path to write the final source to. This is the third
 -- argument passed to the preprocessor.
-newtype Destination = Destination { unDestination :: FilePath }
+newtype Destination = Destination {unDestination :: FilePath}
 
 data AllModelsFile = AllModelsFile
     { amfModuleBase :: Module
     , amfModuleImports :: [Module]
     }
 
+printDebug :: Bool -> String -> IO ()
+printDebug enabled str = when enabled $ putStrLn ("[DEBUG] " ++ str)
+
 -- |
 --
 -- @since 0.1.0.0
 spliceImports
     :: Source
     -> SourceContents
     -> Destination
+    -> Maybe FilePath
+    -> String
+    -> [String]
+    -> Bool
     -> IO ()
-spliceImports (Source src) (SourceContents srcContents) (Destination dest) = do
-    let (dir, file) = splitFileName src
-    files <- filter (/= file) <$> getFilesRecursive dir
+spliceImports (Source src) (SourceContents srcContents) (Destination dest) msearchDir pat prefixes debug = do
+    let
+        (sourceDir, _file) = splitFileName src
+        searchDir = fromMaybe sourceDir msearchDir
+        excludePrefixFilter target = not $ any (`isPrefixOf` target) prefixes
+
+    printDebug debug $ "searching directory: " ++ searchDir
+    printDebug debug $ "searching with pattern: " ++ pat
+    printDebug debug $ "excluding file name: " ++ src
+    eitherFiles <- fmap (filter (essentiallyDistinct src)) <$> getFiles searchDir pat
+    files <- case eitherFiles of
+        Left e -> error e
+        Right f -> pure f
+    let
+        filteredFiles = filter excludePrefixFilter files
+    printDebug debug $ "including files:\n" ++ intercalate "\n|  " filteredFiles
     let
         input =
             AllModelsFile
                 { amfModuleBase =
                     fromJust $ pathToModule src
                 , amfModuleImports =
-                    mapMaybe pathToModule (fmap (dir </>) files)
+                    mapMaybe pathToModule (fmap (searchDir </>) filteredFiles)
                 }
         output =
             renderFile input srcContents
 
     writeFile dest output
+    where
+      essentiallyDistinct :: FilePath -> FilePath -> Bool
+      essentiallyDistinct l r = simplifyPath (Text.pack l) /= simplifyPath (Text.pack r)
+
+      simplifyPath :: Text -> Text
+      simplifyPath str =
+        let
+          intermediate = Text.replace "//" "/" str
+        in
+          if intermediate == str
+          then intermediate
+          else simplifyPath intermediate
```

Comment:
You can do this in one pass - consider `Text.splitOn`, `Text.intercalate`, and `Data.List.group`

Reply:
ah, `group` is what I was missing